beacon-chain/sync/backfill/batch.go

```diff
@@ -0,0 +1,148 @@
+package backfill
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	eth "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	log "github.com/sirupsen/logrus"
+)
+
+var ErrChainBroken = errors.New("batch is not the ancestor of backfilled batch")
+
+type batchState int
+
+func (s batchState) String() string {
+	switch s {
+	case batchNil:
+		return "nil"
+	case batchInit:
+		return "init"
+	case batchSequenced:
+		return "sequenced"
+	case batchErrRetryable:
+		return "error_retryable"
+	case batchImportable:
+		return "importable"
+	case batchImportComplete:
+		return "import_complete"
+	case batchEndSequence:
+		return "end_sequence"
+	default:
+		return "unknown"
+	}
+}
+
+const (
+	batchNil batchState = iota
+	batchInit
+	batchSequenced
+	batchErrRetryable
+	batchImportable
+	batchImportComplete
+	batchEndSequence
+)
+
+type batchId string
+
+type batch struct {
+	firstScheduled time.Time
+	scheduled      time.Time
+	seq            int // sequence identifier, ie how many times has the sequence() method served this batch
+	retries        int
+	begin          primitives.Slot
+	end            primitives.Slot // half-open interval, [begin, end), ie >= start, < end.
+	results        VerifiedROBlocks
+	err            error
+	state          batchState
+	pid            peer.ID
+}
+
+func (b batch) logFields() log.Fields {
+	return map[string]interface{}{
+		"batch_id":  b.id(),
+		"state":     b.state.String(),
+		"scheduled": b.scheduled.String(),
+		"seq":       b.seq,
+		"retries":   b.retries,
+		"begin":     b.begin,
+		"end":       b.end,
+		"pid":       b.pid,
+	}
+}
+
+func (b *batch) inc() {
+	b.seq += 1
+}
+
+func (b batch) replaces(r batch) bool {
+	if b.begin != r.begin {
+		return false
+	}
+	if b.end != r.end {
+		return false
+	}
+	return b.seq >= r.seq
```

Comment:
Should we worry about misuse where we replace a batch that has `batchImportComplete` with something that doesn't?

Specifically if we send this on `update` below coud cause trouble.

Reply:
I can add a check that always returns false if `r` has state `batchImportComplete`.